### PR TITLE
CI: Query Engine tests on Vitess

### PR DESCRIPTION
**Pull request by pimeys**
The last step of the CI test setup for Vitess: Query Engine. Tests run, again, against vitess in single-threaded mode, on GitHub Actions only.

Closes: https://github.com/prisma/prisma-engines/issues/1851
Closes: https://github.com/prisma/prisma/issues/5219
Closes: https://github.com/prisma/migrations-team/issues/203

### CODE REVIEW

File: query-engine/connector-test-kit/src/test/scala/binary/QueryEngine.scala [hunk 1]
```diff
@@ -4,7 +4,7 @@ import org.scalatest.{FlatSpec, Matchers}
 import util._
 
 class QueryEngine extends FlatSpec with Matchers with ApiSpecBase {
-  "Setting a data source override via env" should "prevent env errors" in {
+  "Setting a data source override via env" should "prevent env errors" taggedAs (IgnoreVitess) in {
```
Comment (janpio) [inline query-engine/connector-test-kit/src/test/scala/binary/QueryEngine.scala:7]: TODO

File: query-engine/connector-test-kit/src/test/scala/util/Project.scala [hunk 2]
```diff
@@ -19,9 +19,14 @@ case class Project(
...
     val config = ConnectorConfig.instance
 
+    val provider = config.provider.stripSuffix("56") match {
+      case "vitess" => "mysql"
+      case provider => provider
+    }
+
     s"""
            |datasource test {
-           |  provider = "${config.provider.stripSuffix("56")}"
```
Comment (janpio) [inline query-engine/connector-test-kit/src/test/scala/util/Project.scala:24]: Huh why does this stripping exist?

File: query-engine/connector-test-kit/src/test/scala/util/TestServer.scala [hunk 3]
```diff
@@ -19,7 +19,7 @@ case class TestServer() extends PlayJsonExtensions with LogSupport {
       project: Project,
       dataContains: String = "",
       legacy: Boolean = true,
-      batchSize: Int = 5000,
+      batchSize: Int = 500,
```
Comment (janpio) [inline query-engine/connector-test-kit/src/test/scala/util/TestServer.scala:22]: TODO
Introduces uncertainty about previous value I guess?

File: query-engine/connector-test-kit/src/test/scala/writes/ids/BringYourOwnIdSpec.scala [hunk 4]
```diff
@@ -23,8 +23,9 @@ class BringYourOwnIdSpec extends FlatSpec with Matchers with ApiSpecBase with Sc
       res.toString should be(s"""{"data":{"createParent":{"p":"Parent","id":"Own Id"}}}""")
 
       val errorTarget = () match {
-        case _ if connectorTag == ConnectorTag.MySqlConnectorTag => "constraint: `PRIMARY`"
-        case _                                                   => "fields: (`id`)"
+        case _ if connectorTag == ConnectorTag.MySqlConnectorTag    => "constraint: `PRIMARY`"
+        case _ if connectorTag == ConnectorTag.VitessConnectorTag   => "(not available)"
```
Comment (janpio) [inline query-engine/connector-test-kit/src/test/scala/writes/ids/BringYourOwnIdSpec.scala:27]: TODO

File: query-engine/connector-test-kit/src/test/scala/writes/ids/BringYourOwnIdSpec.scala [hunk 5]
```diff
@@ -101,8 +102,9 @@ class BringYourOwnIdSpec extends FlatSpec with Matchers with ApiSpecBase with Sc
       res.toString should be(s"""{"data":{"createParent":{"p":"Parent","id":"Own Id","childOpt":{"c":"Child","id":"Own Child Id"}}}}""")
 
       val constraintTarget = () match {
-        case _ if connectorTag == ConnectorTag.MySqlConnectorTag => "constraint: `PRIMARY`"
-        case _                                                   => "fields: (`id`)"
+        case _ if connectorTag == ConnectorTag.MySqlConnectorTag  => "constraint: `PRIMARY`"
+        case _ if connectorTag == ConnectorTag.VitessConnectorTag => "(not available)"
```
Comment (janpio) [inline query-engine/connector-test-kit/src/test/scala/writes/ids/BringYourOwnIdSpec.scala:106]: TODO

File: query-engine/connectors/sql-query-connector/src/query_builder/write.rs [hunk 6]
```diff
@@ -155,10 +156,16 @@ pub fn create_relation_table_records(
...
-    let columns: Vec<_> = parent_columns.into_iter().chain(child_columns).collect();
+    let parent_columns = field
+        .related_field()
+        .m2m_columns()
+        .into_iter()
+        .map(|c| c.name.into_owned());
+
+    let child_columns = field.m2m_columns().into_iter().map(|c| c.name.into_owned());
+
+    let columns: Vec<_> = parent_columns.chain(child_columns).collect();
```
Comment (janpio) [inline query-engine/connectors/sql-query-connector/src/query_builder/write.rs:168]: These changes here above are just different code, no real difference in logic, right?

**Comment by janpio** [comment]
<3

File: query-engine/connector-test-kit/src/test/scala/util/TestServer.scala (hunk 3 above)
Comment (pimeys) [reply query-engine/connector-test-kit/src/test/scala/util/TestServer.scala:22]: We don't need that, I just disabled the VeryMany tests anyhow...

File: query-engine/connectors/sql-query-connector/src/query_builder/write.rs (hunk 6 above)
Comment (pimeys) [reply query-engine/connectors/sql-query-connector/src/query_builder/write.rs:168]: I need to backtrack these. We do it in Quaint. This breaks MSSQL.

File: query-engine/connector-test-kit/src/test/scala/util/Project.scala (hunk 2 above)
Comment (pimeys) [reply query-engine/connector-test-kit/src/test/scala/util/Project.scala:24]: It's legacy stuff. Don't really need to know...

File: Makefile [hunk 7]
```diff
@@ -120,9 +120,17 @@ dev-mongodb: start-mongodb
...
 
+dev-vitess_5_7: start-vitess_5_7
+	echo 'vitess_5_7' > current_connector
+	cp $(CONFIG_PATH)/vitess_5_7 $(CONFIG_FILE)
+
 start-vitess_8_0:
 	docker-compose -f docker-compose.yml up -d --remove-orphans vitess-test-8_0 vitess-shadow-8_0
 
+dev-vitess_8_0: start-vitess_8_0
+	echo 'vitess_5_7' > current_connector
```
Comment (janpio) [inline Makefile:131]:
```suggestion
	echo 'vitess_8_0' > current_connector
```

Comment (pimeys) [reply Makefile]: yikes.

File: migration-engine/core/src/qe_setup.rs [hunk 8]
```diff
@@ -0,0 +1,72 @@
+//! Query Engine test setup.
+
+use crate::api::GenericApi;
+use crate::commands::SchemaPushInput;
+use crate::core_error::{CoreError, CoreResult};
+#[cfg(feature = "mongodb")]
+use datamodel::common::provider_names::MONGODB_SOURCE_NAME;
+use datamodel::common::provider_names::{
+    MSSQL_SOURCE_NAME, MYSQL_SOURCE_NAME, POSTGRES_SOURCE_NAME, SQLITE_SOURCE_NAME,
+};
```
Comment (janpio) [inline migration-engine/core/src/qe_setup.rs:10]: nit: Is there a reason these are different?

File: query-engine/connector-test-kit-rs/query-tests-setup/src/connector_tag/vitess.rs [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        })
+    }
+
+    fn capabilities(&self) -> &[ConnectorCapability] {
+        &self.capabilities
+    }
+
+    fn as_parse_pair(&self) -> (String, Option<String>) {
+        let version = self.version.as_ref().map(ToString::to_string);
+        ("sqlite".to_owned(), version)
```
Comment (janpio) [inline query-engine/connector-test-kit-rs/query-tests-setup/src/connector_tag/vitess.rs:38]: This should probably be different?

**Review by janpio — APPROVED** [review]

**Pull request merged**